### PR TITLE
Refactor booking section logic

### LINKED ISSUE #544: FAQs are not loading the correct faqs categories for the booking status
Repro steps:

- user uses iPhone or other mobile device to access SmartFAQ
- user logs into SmartFAQ

Expected
- list of FAQs is updated according to booking status

Actual
- list of FAQs is not updated until the booking info is clicked & expanded

Pls add cypress tests for this or adjust the existing one.

### PR DESCRIPTION
**Pull request by kantord**
fixes #544

**Notes for testers, code review and myself**
As this refactor and fix touches a bunch of components, we have to be more thorough in testing. The following things have to be taken into account:

- Has to work on mobile and desktop (:heavy_check_mark: by @kantord) :hourglass: 
- Has to work for all booking types (oneway, return, multicity)  (:heavy_check_mark: by @kantord) :hourglass: 
- Has to work for nearest booking and manually selected booking  (:heavy_check_mark: by @kantord) :hourglass: 
- Has to work for all login methods including simple token (email link)  :hourglass: :hourglass: 
- *code reviewers*: it cannot make redundant graphql queries  :hourglass: :hourglass: 
- has to fix the original issue obviously :-) (:heavy_check_mark: by @kantord) :hourglass: 
- logout has to work, should show correct category after logout (:heavy_check_mark: by @kantord) :hourglass: 
- no tests should be deleted (:heavy_check_mark: by @kantord) :hourglass: 
- *code reviewers*: we have to test that it doesn't break Baggage (disabled feature)  :hourglass: :hourglass: 
- non-airplane vehicles still work  :hourglass: :hourglass: 
- no queries left in SingleBookingPage or MobileBookingHeader (:heavy_check_mark: : by @kantord) :hourglass: 
- everything reacts correctly to network errors  :hourglass: :hourglass: 

This technically doesn't have to be resolved here, since it's an unrelated issue already present in production, but it'd be nice to solve it here, or create a new issue:
- until booking data is not loaded, no faqs should be loaded (in other terms, only one query is made)   (:x: by @kantord) :hourglass:

### CODE REVIEW

File: src/context/BookingState.js [hunk 1]
```diff
@@ -98,14 +102,16 @@ class BookingStateProvider extends React.Component<Props, BookingStateType> {
   }
 }
 
-export const withLogout = <Props>(Component: React.ComponentType<{} & Props>) =>
+export const withLogout = <Props>(
```
Comment (jaroslav-kubicek) [inline src/context/BookingState.js]: this definitely doesn't belong into this file

File: src/context/ActiveTab.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    const activateTab = (tab: TabType) => {
+      this.setState({ activeTab: tab });
+    };
+
+    const toggleTab = (tab: TabType) => {
+      this.setState(prevState => ({
+        activeTab: prevState.activeTab === tab ? null : tab,
+      }));
+      this.forceUpdate();
```
Comment (jalbertsr) [inline src/context/ActiveTab.js]: we should try to avoid all uses of `forceUpdate()`

Comment (kantord) [reply src/context/ActiveTab.js]: Agreed, but I didn't add that, I just moved it from another part of the code. We should remove it, but maybe this PR is big enough already :-)

File: src/context/BookingState.js (hunk 1 above)
Comment (kantord) [reply src/context/BookingState.js]: agreed

**Comment by kantord** [comment]
> Could we do such refactorings while keeping relay fragments? I don't believe this is a optimal way to go.

Do you think it's possible using context?

File: src/BookingStateWrapper/MultiCityTrip.js [hunk 3]
```diff
@@ -14,8 +15,8 @@ const MultiCityTrip = ({ children, booking }: Props) => (
   <BookingStateProvider
     hasBooking
     isPastBooking={booking.isPastBooking}
-    isUrgent={false /*fixme*/}
-    onLogout={async () => null}
+    departureTime={idx(booking, _ => _.trips[0].departure.time)}
+    booking={{ props: booking, error: null }}
```
[the remark this answers is not in the document]
Comment (kantord) [reply src/BookingStateWrapper/MultiCityTrip.js]: Do you mean `null`?

[the remark this answers is not in the document]
Comment (jaroslav-kubicek) [reply src/BookingStateWrapper/MultiCityTrip.js]: 🤕  yes

**Comment by jaroslav-kubicek** [comment]
> Do you think it's possible using context?

Yes I believe, child components don't need to work with context, they should instead declare all their dependencies though fragments. Context should be needed only for "root" component working with booking which renders these child components 🙂

**Comment by kantord** [comment]
closing in favor of another iteration :-) https://github.com/kiwicom/smart-faq/pull/715

**Pull request closed without merging**